addons/mail/models/res_partner.py

```diff
@@ -126,16 +126,25 @@ def get_static_mention_suggestions(self):
         return [partner.mail_partner_format() for partner in partners]
 
     @api.model
-    def get_mention_suggestions(self, search, limit=8, channel_id=None):
+    def get_mention_suggestions(self, search, limit=8):
         """ Return 'limit'-first partners' such that the name or email matches a 'search' string.
             Prioritize partners that are also users, and then extend the research to all partners.
-            If channel_id is given, only members of this channel are returned.
             The return format is a list of partner data (as per returned by `mail_partner_format()`).
         """
+        partners = self._get_mention_suggestions(search, limit)
+        return [partner.mail_partner_format() for partner in partners]
+
+    @api.model
+    def get_channel_mention_suggestions(self, search, limit=8, channel_id=None):
+        partners = self._get_mention_suggestions(search, limit)
+        partner_infos = {partner.id: partner.mail_partner_format() for partner in partners}
+        members = self.env['mail.channel.partner'].search([('channel_id', '=', channel_id), ('partner_id', 'in', partners.ids)])
+        return members.mail_channel_partner_format(partner_infos)
+
+    @api.model
+    def _get_mention_suggestions(self, search, limit):
         search_dom = expression.OR([[('name', 'ilike', search)], [('email', 'ilike', search)]])
         search_dom = expression.AND([[('active', '=', True), ('type', '!=', 'private')], search_dom])
-        if channel_id:
-            search_dom = expression.AND([[('channel_ids', 'in', channel_id)], search_dom])
```

Comment:
domain should be kept inside, if you post-filter it then the limit won't work as expected